compiler-core/src/io.rs

```diff
@@ -56,6 +55,63 @@ pub struct OutputFile {
     pub path: PathBuf,
 }
 
+#[allow(missing_debug_implementations)]
+pub struct ReadDir {
+    entries: Box<dyn Iterator<Item = io::Result<DirEntry>>>,
+}
+
+impl ReadDir {
+    pub fn from_entries(entries: Vec<io::Result<DirEntry>>) -> Self {
```

Comment:
Could this function take `Box<dyn Iterator<Item = io::Result<DirEntry>>>`? Seems we build vecs from iterators to turn them back to iterators, that seems like it could be avoided.

Reply:
I tried this and unfortunately it leads to lifetime issues with the closures involved with the boxed `dyn Iterator` which leads me to believe we need the allocation to avoid that issue. I was able to unbox the entries field and updated `ReadDir` to work with the `FromIterator` and `IntoIterator` traits which cleans up the code a bit... I think? 2ab5e588